### PR TITLE
sql: allow stars inside view definitions

### LINKED ISSUE #10028: sql: Support view queries with star expansions
Copied over from https://github.com/cockroachdb/cockroach/issues/9921#issuecomment-254009800

We currently don't expand out star selectors before persisting a new view's query. If a dependent table has columns added to it, that can break later queries from the view depending on how that dependent table was used.

It's kind of a fluke, but a simple star select works just fine as columns are added because the view descriptor only lists the original columns as its ResultColumns:

```
root@:26257> create table star (a int primary key, b int);
CREATE TABLE
root@:26257> create view v as select * from star;
CREATE VIEW
root@:26257> insert into star values (1, 1), (2, 2);
INSERT 2
root@:26257> select * from star;
+---+---+
| a | b |
+---+---+
| 1 | 1 |
| 2 | 2 |
+---+---+
(2 rows)
root@:26257> SELECT * FROM v;
+---+---+
| a | b |
+---+---+
| 1 | 1 |
| 2 | 2 |
+---+---+
(2 rows)
root@:26257> alter table star add column c int;
ALTER TABLE
root@:26257> SELECT * FROM star;
+---+---+------+
| a | b |  c   |
+---+---+------+
| 1 | 1 | NULL |
| 2 | 2 | NULL |
+---+---+------+
(2 rows)
root@:26257> SELECT * FROM v;
+---+---+
| a | b |
+---+---+
| 1 | 1 |
| 2 | 2 |
+---+---+
(2 rows)
```

On the other hand, if the \* isn't at the end of the query, it can break the view as columns are added:

```
root@:26257> create table t (a int primary key, b int);
CREATE TABLE
root@:26257> insert into t values (1, 1), (2, 2);
INSERT 2
root@:26257> CREATE VIEW tv AS SELECT t1.*, t2.b AS t2b FROM t AS t1 JOIN t AS t2 ON t1.a = t2.a;
CREATE VIEW
root@:26257> select * from tv;
+---+---+-----+
| a | b | t2b |
+---+---+-----+
| 1 | 1 |   1 |
| 2 | 2 |   2 |
+---+---+-----+
(2 rows)
root@:26257> ALTER TABLE t ADD COLUMN c INT;
ALTER TABLE
root@:26257> SELECT t1.*, t2.b AS t2b FROM t AS t1 JOIN t AS t2 ON t1.a = t2.a;
+---+---+------+-----+
| a | b |  c   | t2b |
+---+---+------+-----+
| 1 | 1 | NULL |   1 |
| 2 | 2 | NULL |   2 |
+---+---+------+-----+
(2 rows)
root@:26257> SELECT * FROM tv;
+---+---+------+
| a | b | t2b  |
+---+---+------+
| 1 | 1 | NULL |
| 2 | 2 | NULL |
+---+---+------+
(2 rows)
```

The logic for expanding stars appears to be a little broader in scope than the logic for normalizing table names, so I'm not sure if there's as well-contained of a fix for it as for #9921 (beyond moving to a semantic representation, of course).

Is this the sort of thing that we should block the rollout of views for? Is it ok to leave as a known issue? Is there a fix that could reasonably be put in place for it?

@knz @dt @nvanbenschoten 

Epic CRDB-2410

Jira issue: CRDB-6144

### PR DESCRIPTION
**Pull request by knz**
First commit from #97544.
Fixes #10028.

Release note (sql change): It is now possible to use `*` inside CREATE VIEW. The list of columns is expanded at the time the view is created, so that new columns added after the view was defined are not included in the view. This behavior is the same as PostgreSQL.

### CODE REVIEW

File: pkg/sql/logictest/testdata/logic_test/views [hunk 1]
```diff
@@ -380,68 +380,170 @@ CREATE VIEW virt2 AS SELECT range_id, lease_holder FROM crdb_internal.ranges
...
+          FROM
+            test.public.t AS t1
+            JOIN test.public.t AS t2 ON t1.a IN (SELECT a FROM (SELECT t.a, t.b FROM test.public.t))
+star2   CREATE VIEW public.star2 (
+          a,
+          b
+        ) AS SELECT t.a, t.b FROM test.public.t
+star3   CREATE VIEW public.star3 (
+          a
+        ) AS SELECT a FROM test.public.t ORDER BY t.*
```
Comment (knz) [inline pkg/sql/logictest/testdata/logic_test/views:486]: I am noting how the stars inside ORDER BY and GROUP BY are preserved here. This means they are also perserved in UDFs.

I believe this is OK? But I think we need some semantic expert to double check. In any case, this choice is not specific to this PR so it should not block it.

**Review by DrewKimball — COMMENTED** [review]
> Reviewed 1 of 4 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/97515)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @knz, @mgartner, and @rharding6373)

___
*[`pkg/sql/opt/optbuilder/project.go` line 151 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/97515#-NOuyF-M3AR9y9t8CNw-:-NOuyF-M3AR9y9t8CNw0:b-csdgmr) ([raw file](https://github.com/cockroachdb/cockroach/blob/09d2c5df77c9fef1b0fb6bb277c54e72facbca78/pkg/sql/opt/optbuilder/project.go#L151)):*
> ```Go
> 
> 					aliases, exprs := b.expandStar(e.Expr, inScope)
> 					if b.insideFuncDef || b.insideViewDef {
> ```

I think there's another case in `Builder.buildDataSource` that needs to be changed (see #96375).
___
*[`pkg/sql/logictest/testdata/logic_test/views` line 439 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/97515#-NOuzigO0_wLa00CewJt:-NOuzigO0_wLa00CewJu:b-jwituc) ([raw file](https://github.com/cockroachdb/cockroach/blob/09d2c5df77c9fef1b0fb6bb277c54e72facbca78/pkg/sql/logictest/testdata/logic_test/views#L439)):*
> ```
> 
> statement ok
> CREATE VIEW star16 AS SELECT t1.a, t2.a AS a2 FROM t AS t1 JOIN t AS t2 ON t1.a IN (SELECT a FROM (SELECT * FROM t))
> ```

Could you also add the cases from #96375? (similar to the existing join test cases, but without aliases)
___
*[`pkg/sql/logictest/testdata/logic_test/views` line 486 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/97515#-NOus2K3iUKFUQkqNrPs-r1-486:-NOv-RTqCh4V83XhtvW6:b81o75) ([raw file](https://github.com/cockroachdb/cockroach/blob/09d2c5df77c9fef1b0fb6bb277c54e72facbca78/pkg/sql/logictest/testdata/logic_test/views#L486)):*
<details><summary><i>Previously, knz (Raphael 'kena' Poss) wrote…</i></summary><blockquote>

I am noting how the stars inside ORDER BY and GROUP BY are preserved here. This means they are also perserved in UDFs.

I believe this is OK? But I think we need some semantic expert to double check. In any case, this choice is not specific to this PR so it should not block it.
</blockquote></details>

That's a good catch, and I think it's a bug.

**Review by knz — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/97515)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @DrewKimball, @mgartner, and @rharding6373)

___
*[`pkg/sql/opt/optbuilder/project.go` line 151 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/97515#-NOuyF-M3AR9y9t8CNw-:-NOv1R3n7IFJ-p9bam8R:bkhzaye) ([raw file](https://github.com/cockroachdb/cockroach/blob/09d2c5df77c9fef1b0fb6bb277c54e72facbca78/pkg/sql/opt/optbuilder/project.go#L151)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

I think there's another case in `Builder.buildDataSource` that needs to be changed (see #96375).
</blockquote></details>

hum no that would be a regression - we currently allow views to reference subqueries withut an alias, so removing that here would be a regression from current behavior.
___
*[`pkg/sql/logictest/testdata/logic_test/views` line 439 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/97515#-NOuzigO0_wLa00CewJt:-NOv1jjB-SrsvVBUqZX9:b-zii22h) ([raw file](https://github.com/cockroachdb/cockroach/blob/09d2c5df77c9fef1b0fb6bb277c54e72facbca78/pkg/sql/logictest/testdata/logic_test/views#L439)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

Could you also add the cases from #96375? (similar to the existing join test cases, but without aliases)
</blockquote></details>

i'll look into it.
___
*[`pkg/sql/logictest/testdata/logic_test/views` line 486 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/97515#-NOus2K3iUKFUQkqNrPs-r1-486:-NOv1hmtDUoppqbGqPv7:b1p62rl) ([raw file](https://github.com/cockroachdb/cockroach/blob/09d2c5df77c9fef1b0fb6bb277c54e72facbca78/pkg/sql/logictest/testdata/logic_test/views#L486)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

That's a good catch, and I think it's a bug.
</blockquote></details>

i'll file an issue.

**Review by knz — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/97515)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @DrewKimball, @mgartner, and @rharding6373)

___
*[`pkg/sql/logictest/testdata/logic_test/views` line 486 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/97515#-NOus2K3iUKFUQkqNrPs-r1-486:-NOv2zlD2GywOB4dQzHj:b8v3wkp) ([raw file](https://github.com/cockroachdb/cockroach/blob/09d2c5df77c9fef1b0fb6bb277c54e72facbca78/pkg/sql/logictest/testdata/logic_test/views#L486)):*
<details><summary><i>Previously, knz (Raphael 'kena' Poss) wrote…</i></summary><blockquote>

i'll file an issue.
</blockquote></details>

#97520

**Review by knz — COMMENTED** [review]
RFAL

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/97515#-:-NOv54Sp7OtlYuR-LySf:bb72pd0)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @DrewKimball, @mgartner, and @rharding6373)

___
*[`pkg/sql/opt/optbuilder/project.go` line 151 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/97515#-NOuyF-M3AR9y9t8CNw-:-NOv4skHEIJUaWIEvbkb:b-auk9rp) ([raw file](https://github.com/cockroachdb/cockroach/blob/09d2c5df77c9fef1b0fb6bb277c54e72facbca78/pkg/sql/opt/optbuilder/project.go#L151)):*
<details><summary><i>Previously, knz (Raphael 'kena' Poss) wrote…</i></summary><blockquote>

hum no that would be a regression - we currently allow views to reference subqueries withut an alias, so removing that here would be a regression from current behavior.
</blockquote></details>

I double checked - we do not need the special case related to #96375 here because CREATE VIEW separately rejects duplicate column names, so if there's an ambiguity it will be prevented this way.
___
*[`pkg/sql/logictest/testdata/logic_test/views` line 439 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/97515#-NOuzigO0_wLa00CewJt:-NOv4rti6KA4xIQCQWpC:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/09d2c5df77c9fef1b0fb6bb277c54e72facbca78/pkg/sql/logictest/testdata/logic_test/views#L439)):*
<details><summary><i>Previously, knz (Raphael 'kena' Poss) wrote…</i></summary><blockquote>

i'll look into it.
</blockquote></details>

Done.

**Review by rharding6373 — APPROVED** [review]
Very nice, thanks for all the test cases! I only have suggestions for a couple more. :lgtm: 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/97515#-:-NOv9jsI5N13kx585enJ:bvpogh5)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @ajwerner, @DrewKimball, @knz, and @mgartner)

___
*[`pkg/sql/logictest/testdata/logic_test/views` line 440 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/97515#-NOv7tcnBgAQaPDsMmjx:-NOv7tcnBgAQaPDsMmjy:b-wt6nr3) ([raw file](https://github.com/cockroachdb/cockroach/blob/09d2c5df77c9fef1b0fb6bb277c54e72facbca78/pkg/sql/logictest/testdata/logic_test/views#L440)):*
> ```
> statement ok
> CREATE VIEW star16 AS SELECT t1.a, t2.a AS a2 FROM t AS t1 JOIN t AS t2 ON t1.a IN (SELECT a FROM (SELECT * FROM t))
> ```

Could you please add the following tests (or similar) to make sure schema changes and dependencies are handled correctly? 

```
statement ok
ALTER TABLE t ADD COLUMN c INT;

statement error
ALTER TABLE t DROP COLUMN b;

# we expect the columns in the view to be unchanged
statement ok
ALTER TABLE t RENAME COLUMN b TO d;
```

**Review by DrewKimball — APPROVED** [review]
:lgtm:

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/97515#-:-NOvBk3QCjJhvGICi-xf:bnfp4nl)* status: :shipit: complete! 2 of 0 LGTMs obtained (waiting on @ajwerner, @knz, and @mgartner)

___
*[`pkg/sql/opt/optbuilder/project.go` line 151 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/97515#-NOuyF-M3AR9y9t8CNw-:-NOv77wX6CCcjWtDjz00:ba7duo0) ([raw file](https://github.com/cockroachdb/cockroach/blob/09d2c5df77c9fef1b0fb6bb277c54e72facbca78/pkg/sql/opt/optbuilder/project.go#L151)):*
>CREATE VIEW separately rejects duplicate column names

I think this will only apply to the top-level columns returned by the view, but it seems ok anyway since apparently views bind differently from UDFs: https://github.com/cockroachdb/cockroach/issues/96375#issuecomment-1440882770
___
*[`pkg/sql/logictest/testdata/logic_test/views` line 468 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/97515#-NOv7Z5EA3MoUzLxAGWl:-NOv7Z5EA3MoUzLxAGWm:b-17u3rs) ([raw file](https://github.com/cockroachdb/cockroach/blob/8298d85cc8ab00ea363c7d3cc8ea82941db89deb/pkg/sql/logictest/testdata/logic_test/views#L468)):*
> ```
> 
> 
> # This error is the same as in postgres.
> ```

[nit] postgres gives `subquery in FROM must have an alias`, so not the same error here.

**Review by knz — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/97515)* status: :shipit: complete! 2 of 0 LGTMs obtained (waiting on @ajwerner, @DrewKimball, @mgartner, and @rharding6373)

___
*[`pkg/sql/opt/optbuilder/project.go` line 151 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/97515#-NOuyF-M3AR9y9t8CNw-:-NOwMtp8En3uHFbNgUFd:b3tcbt7) ([raw file](https://github.com/cockroachdb/cockroach/blob/09d2c5df77c9fef1b0fb6bb277c54e72facbca78/pkg/sql/opt/optbuilder/project.go#L151)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

>CREATE VIEW separately rejects duplicate column names

I think this will only apply to the top-level columns returned by the view, but it seems ok anyway since apparently views bind differently from UDFs: https://github.com/cockroachdb/cockroach/issues/96375#issuecomment-1440882770
</blockquote></details>

I investigated this further and we have indeed an 'interesting' situation here:
```sql
select array[z.*]::string from (select * from (select a from t), (select a from t)) z;
```

This example is interesting because there's just 1 unambiguous "outer" column, but we need to be careful "in the middle".

With my current code, we get invalid behavior as follows:
```
demo@127.0.0.1:26257/movr> show create view v;
  table_name |                                       create_statement
-------------+-----------------------------------------------------------------------------------------------
  v          | CREATE VIEW public.v (
             |     "array"
             | ) AS SELECT
             |         ARRAY[z.*]::STRING
             |     FROM
             |         (SELECT a, a FROM (SELECT a FROM movr.public.t), (SELECT a FROM movr.public.t)) AS z
(1 row)

Time: 9ms total (execution 9ms / network 0ms)

demo@127.0.0.1:26257/movr> select * from v;
ERROR: column reference "a" is ambiguous (candidates: <anonymous>.a)
SQLSTATE: 42702
```

Which means my PR needs to solve a subset of #96375 before we consider the job done.
___
*[`pkg/sql/logictest/testdata/logic_test/views` line 440 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/97515#-NOv7tcnBgAQaPDsMmjx:-NOwK2wc9IOpe1Kd2eWR:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/09d2c5df77c9fef1b0fb6bb277c54e72facbca78/pkg/sql/logictest/testdata/logic_test/views#L440)):*
<details><summary><i>Previously, rharding6373 (Rachael Harding) wrote…</i></summary><blockquote>

Could you please add the following tests (or similar) to make sure schema changes and dependencies are handled correctly? 

```
statement ok
ALTER TABLE t ADD COLUMN c INT;

statement error
ALTER TABLE t DROP COLUMN b;

# we expect the columns in the view to be unchanged
statement ok
ALTER TABLE t RENAME COLUMN b TO d;
```
</blockquote></details>

Done.
___
*[`pkg/sql/logictest/testdata/logic_test/views` line 468 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/97515#-NOv7Z5EA3MoUzLxAGWl:-NOwKenI93DHNf9A0sRP:bklp7k5) ([raw file](https://github.com/cockroachdb/cockroach/blob/8298d85cc8ab00ea363c7d3cc8ea82941db89deb/pkg/sql/logictest/testdata/logic_test/views#L468)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

[nit] postgres gives `subquery in FROM must have an alias`, so not the same error here.
</blockquote></details>

Good point. Changed to add a test using pg's syntax.

**Review by knz — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/97515)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 2 stale) (waiting on @ajwerner, @DrewKimball, @mgartner, and @rharding6373)

___
*[`pkg/sql/opt/optbuilder/project.go` line 151 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/97515#-NOuyF-M3AR9y9t8CNw-:-NOwVwUwEkylYOYgQQrk:b-d1jtsr) ([raw file](https://github.com/cockroachdb/cockroach/blob/09d2c5df77c9fef1b0fb6bb277c54e72facbca78/pkg/sql/opt/optbuilder/project.go#L151)):*
<details><summary><i>Previously, knz (Raphael 'kena' Poss) wrote…</i></summary><blockquote>

I investigated this further and we have indeed an 'interesting' situation here:
```sql
select array[z.*]::string from (select * from (select a from t), (select a from t)) z;
```

This example is interesting because there's just 1 unambiguous "outer" column, but we need to be careful "in the middle".

With my current code, we get invalid behavior as follows:
```
demo@127.0.0.1:26257/movr> show create view v;
  table_name |                                       create_statement
-------------+-----------------------------------------------------------------------------------------------
  v          | CREATE VIEW public.v (
             |     "array"
             | ) AS SELECT
             |         ARRAY[z.*]::STRING
             |     FROM
             |         (SELECT a, a FROM (SELECT a FROM movr.public.t), (SELECT a FROM movr.public.t)) AS z
(1 row)

Time: 9ms total (execution 9ms / network 0ms)

demo@127.0.0.1:26257/movr> select * from v;
ERROR: column reference "a" is ambiguous (candidates: <anonymous>.a)
SQLSTATE: 42702
```

Which means my PR needs to solve a subset of #96375 before we consider the job done.
</blockquote></details>

Rebased on top of #97544, which fixes this. PTAL.

**Comment by knz** [comment]
TFYRs!

bors r=DrewKimball,rharding6373

**Pull request merged**